### ISSUE TITLE
Language-aware conversion to ASCII

### ISSUE THREAD

**Issue by rolve** [zotero/zotero#791]
Conversion for unicode to ASCII happens in several places in Zotero, including (afaik) for generating Bibtex keys and for naming attachments by ZotFile. This conversion is language-agnostic and just "removes" the diacritics. (I understand that the function `Zotero.Utilities.removeDiacritics` is used for this purpose.)

For some languages (e.g. German), there are more specific rules to convert certain letters to ASCII. For example, `ä` is converted to `ae`, `ö` to `oe`, and `ü` to `ue`.

Therefore, I propose the following improvement. Whenever the language of a string is known, (e.g., for titles of items that have the language field set), language-specific conversion rules like the ones above are taken into account. Otherwise, language-agnostic conversion is performed.

(I imagine a version of `removeDiacritics` with an additional parameter for the language.)

**Comment by fbennett** [comment]
The Python [unidecode](https://pypi.python.org/pypi/Unidecode) package handles this without language hints, and could be adapted. The data set is pretty large, though -- 1.5M. From a quick look at it, maybe half of the volume is made up of comment strings, but it's still quite a large hammer. Maybe @retorquere already handles this in Better BibTeX?

**Comment by retorquere** [comment]
I use a browserified [fold-to-ascii](https://github.com/mplatt/fold-to-ascii), weighing in at 28k.

**Comment by retorquere** [comment]
(which, to be clear, doesn't do language-specific transliteration).

**Comment by fbennett** [comment]
Not sure what the desired scope would be. Should the romanization cover things like Chinese, Korean, Arabic and Hebrew?

**Comment by duncdrum** [comment]
Is this really a good solutions for zotero to dive into language specific ascii conversions? Wouldn't it make more sense for the plugins to either move to unicode, or ti deal with this on their end?

**Comment by fbennett** [comment]
Yeah, it's a slippery slope. Looking at it has me thinking this isn't a task for Zotero itself to take on.

**Comment by retorquere** [comment]
Ah, then there's of course the *other* asciification that BBT does. Not interesting for Zotero of course, but BBT will change `ä` into `\"a` for BibTeX by default. Its not diacritics removal, just mentioning it for completeness.

Why would Zotero want to do "good" transliteration in any case? I can think only of the two cases listed above, and a) generating good BibTeX keys is not really Zotero's core business, and b) how much does it matter for file names?

I would agree that having good citekeys is super important, but that's why BBT allows you to override them; the generated keys are supposed to be sensible defaults.

But what other cases for transliteration exist that aren't covered by this -- except asciification, but for that you don't want transliteration.

**Comment by fbennett** [comment]
I can't think of anything inside Zotero proper that would need refined human-readable transliteration, and the examples given by @rolve are both plugins.

Maybe just putting a comment in the source of `removeDiacritics()` would be enough - so that future plugin authors who come across it will be aware of its limitations, and of other options available.

**Comment by dstillman** [comment]
What about search? Do German users expect to be able to use `ö` and `oe` interchangeably in a search?

**Comment by zuphilip** [comment]
The process of changing `ö` into `oe` is not reversible, because also in German we have words with `oe`: http://www.wordmine.info/de/Search.aspx?slang=de&stype=worter-mit&sword=OE , e.g. "Koeffizient" (=coefficient), but "Köffizient" doesn't make sense. If I want to search for `OECD` I don't want to find results for `ÖCD`. Sometimes it could be nice to search for variants, e.g. `Goethe` and `Göthe` and [all its variants](http://d-nb.info/gnd/118540238). However, in general I wouldn't expect that and it may even be strange to do that. Moreover, for Germans it is no problem to search for words with `ö` because it is part of the German keyboard layout.

**Comment by dstillman** [comment]
OK, thanks — so then for searching it seems we really just want to allow people who might not know how to easily type `ü` to type `gothe` and get a result, in which case `removeDiacritics` would be sufficient.

**Comment by rolve** [comment]
Regarding search, I think having one or two false positives is much better than missing half of an author's publications just because he (or somebody else) didn't spell his name consistently.

**Comment by dstillman** [comment]
Right, I guess that's what I was getting at — the better question is, do we want to allow people to search for `göthe` and see results for both `Göthe` and `Goethe`?

**Comment by rolve** [comment]
In addition, the above list of german words containing proper "oe"s is very short and most words are quite obscure. But the real question is, are there any real "ö" words that might accidentally show up if you searched for the "oe" version of the word? I would guess practically none.

**Comment by dstillman** [comment]
If these rules are language-specific, though, this would get a bit complicated.

For German-tagged items, the normalized word for `göthe` and `goethe` would be `gothe`. We could do the same normalization for searches, so searches for all three would work. `koeffizient` would also be normalized to `koffizient`, but you'd still get the search result you intended.

However, unless we wanted "oe" to find "o" across all items, regardless of language, we'd need to normalize the search term differently for each language in the DB and check each term against the items in those languages.

**Comment by dstillman** [comment]
The other thing is that I sort of doubt people are populating the Language field widely enough that it'd be worth implementing this if it relied on that field. It's one thing for plugins to require the language field for specialized functions like BibTeX keys, but it doesn't seem like we should expect people to do that for standard usage — e.g., searching — in Zotero itself.

**Comment by zuphilip** [comment]
The main question is IMO: do we want an exact search or some kind of fuzzy search? And would there be an option to toggle on/off fuzzy search?

A fuzzy search could do more then just remove diacritics. For example: if I search for "mouse" should I then also see results for "mouses" and "mice"? Or maybe one could allow an edit distance of two between the words. A Solr index would allow some of these tasks, but I guess we don't want to go into that direction.

For searching variants of the author's name, it would be better to use (somehow) external linked data about this person, for example from the GND, Wikidata, or ORCID. 
CC @adam3smith

**Comment by dstillman** [comment]
Once we switch to Electron, we'll have more flexibility in terms of embedding more advanced third-party tools that can handle a lot of this for us (e.g., Solr), so we may not want to bother doing anything until then.

**Comment by fbennett** [comment]
It would be easy to get carried away, but wouldn't it be possible to expand search terms containing a few specific characters (ä, ö, ü, ß, and maybe ) without regard to language, to capture possible variants? That way the expanded search scope would only kick in if the user entered the term with the diacritics, and users who do would likely not be put off by a few false positives.

I looked around for sources on how diacritics are (sometimes) mapped to digraphs, and had no luck until I hit [this comment](http://php.net/manual/en/function.iconv.php#105507) in a discussion thread about `iconv`. It contains locale-specific transliteration tables&mdash;&lsquo;ö&rsquo; in the English locale yields &lsquo;o&rsquo;, but in the German locale it yields &lsquo;oe&rsquo;.

So you could explore a solution for name variants (which I take to be the most common stumbling block) by choosing a list of problematic locales (European languages with roman-like alphabets), collecting their alphabets, and then running all character sets against all locales to see what it kicks out. Any conversions that yield the same result as removeDiacritics can be discarded, as only char -> digraph mappings are of interest. I suspect that you would end up with a small set of mappings that could be used for expanded search.

(While that would be one possible route to explore, I take Dan's point that there may be things out there already that will serve the purpose.)

**Comment by bwiernik** [comment]
There is huge diversity in how databases import diacritic names. Example, I have Paessler, Passler, Päßler, and Pässler, all referring to the same author. As a non-German speaker, I didn't know about these alternatives until I couldn't find an article. For me not knowing language rules, the best experience would be for searches for any of these strings to return all of these names. False positives are definitely better than false negatives. 

A related issue for searches is HTML tags. Currently, if tags are used for rich formatting, they interfere with matching in searches (e.g., I would expect M<sub>2</sub> to be match a search: "M2").
